desktop/src/components/Terminal/useTerminalSearch.tsx

```diff
@@ -0,0 +1,402 @@
+import React, { useCallback, useEffect, useRef, useState } from "react"
+import { TTerminal } from "@/components"
+
+export type TSearchOptions = {
+  searchString?: string
+  caseSensitive?: boolean
+  wholeWordSearch?: boolean
+}
+
+type TSearchResult = [row: number, col: number, len: number]
+
+type TDisplayLine = {
+  index: number
+  text: string
+  inputLine: number
+  startCol: number
+  endCol: number
+}
+
+type THighlight = {
+  displayRow: number
+
+  // Bounds of the highlight within a row.
+  startCol: number
+  endCol: number
+
+  // Which search result this highlight is attached to.
+  resultIndex: number
+}
+
+type TDisplayLineMap = { [key: number]: TDisplayLine[] }
+
+// Used for keeping track of anchors/jump points in the wrapped lines to jump to for any given search result.
+type TJumpMap = { [resultIndex: number]: number }
+
+type TSearchState = {
+  searchableLines: string[]
+  disposables: VoidFunction[]
+  searchOptions: TSearchOptions | undefined
+  jumpMap: TJumpMap
+  debounce: number | undefined
+  activeSearchResult: number
+  preWrappedLines: TDisplayLineMap | undefined
+  searchResults: TSearchResult[]
+  highlights: THighlight[]
+}
+
+export function useTerminalSearch(
+  terminalRef: React.MutableRefObject<TTerminal | null>,
+  searchOptions?: TSearchOptions
+) {
+  const [totalSearchResults, setTotalSearchResults] = useState<number>(-1)
+  const [activeSearchResult, setActiveSearchResult] = useState<number>(0)
+
+  // We have to exercise caution not to re-generate connectStream,
+  // so we have to store a lot of state outside the usual mechanisms.
+  // Otherwise, we will make the terminal flicker.
+
+  const searchStateRef = useRef<TSearchState>({
+    searchableLines: [],
+    disposables: [],
+    searchOptions,
+    jumpMap: {},
+    debounce: undefined,
+    activeSearchResult: 0,
+    preWrappedLines: undefined,
+    searchResults: [],
+    highlights: [],
+  })
+
+  const clearDisposables = useCallback(() => {
+    const toClear = searchStateRef.current.disposables
+    searchStateRef.current.disposables = []
```

Comment:
Could you swap this line with 74 please? This works because we're assigning a new array to `disposables` but I tripped over it when reading and had to think for a moment

Reply:
Done